### PR TITLE
Support list_models and list_datasets with token arg

### LINKED ISSUE #634: List all datasets or models using token
As of now, `list_models` and `list_datasets` don't use the token, so they only list public models. If we could pass the token here, it would be very useful to do things such as retrieving all datasets (including private) of an organization.

cc @muellerzr @LysandreJik 

https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py#L511

### PR DESCRIPTION
**Pull request by muellerzr**
Closes #634

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Yes, LGTM! Thanks @muellerzr.

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -592,9 +593,19 @@ def list_models(
                 This is set to `True` by default when using a filter.
             fetch_config (:obj:`bool`, `optional`):
                 Whether to fetch the model configs as well. This is not included in `full` due to its size.
-
```
Comment (osanseviero) [inline src/huggingface_hub/hf_api.py:595]: nit: revert

**Comment by muellerzr** [comment]
cc @albertvillanova so it's on your radar :)

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -764,9 +776,19 @@ def list_datasets(
...
+                a valid `auth_token` can be passed in as a string.
         """
         path = f"{self.endpoint}/api/datasets"
+        if use_auth_token:
+            if isinstance(use_auth_token, str):
+                if not self._is_valid_token(use_auth_token):
+                    raise ValueError("Invalid token passed!")
+                token = use_auth_token
+            else:
+                token = HfFolder.get_token()
```
Comment (osanseviero) [inline src/huggingface_hub/hf_api.py:790]: Consider something like this for consistency

```suggestion
                token = HfFolder.get_token()
                if token is None:
                  raise EnvironmentError(
                    "You need to provide a `token` or be logged in to Hugging Face with "
                    "`huggingface-cli login`."
                  )
```

**Review by osanseviero — APPROVED** [review]
Looks good! Thanks a lot!

File: src/huggingface_hub/hf_api.py (hunk 2 above)
Comment (muellerzr) [reply src/huggingface_hub/hf_api.py:795]: Opted for this error as it's a bit more specific towards what's happening:
```python
                        "You need to provide a `token` to `use_auth_token` or be logged in to Hugging Face with "
                        "`huggingface-cli login`."
```

Comment (osanseviero) [reply src/huggingface_hub/hf_api.py:795]: You're right

File: src/huggingface_hub/hf_api.py (hunk 1 above)
Comment (osanseviero) [reply src/huggingface_hub/hf_api.py:595]: ignore!

**Pull request merged**